### PR TITLE
Toolbox drag & drop functionality

### PR DESCRIPTION
**Pull request by MushkyQT**
In this PR, I added drag & drop functionality to the Toolbox component.
Expected behavior is as follows:
- When dropping element from Editor component onto Toolbox, cancel return to position animation & delete this element 
- While dragging element from Toolbox, insert placeholder copy of the element in original source position (give "copy" effect)
- When dropping element from Toolbox onto Toolbox, ignore and cancel return to position animation
- When dropping element from Toolbox onto Editor component, insert dropped tag at drop position

Known issues and next set of todos:
- [ ]  [BUG] When element is dropped after last tag, it disappears
- [x]  [BUG] Horizontal width issues while dragging, especially on mobile. Might be placeholder related
Smjugy9YMd

- [x]  [BUG] When moving element from a single-element line, drag behavior issues
cUX4bAS93g

- [ ]  [VISUAL BUG] Toolbox category max-height calculation unexpected behavior on description toggles
- [x]  [6cfff2cdfe1c0f80aad4e39f549a0f7eb3486978] [VISUAL BUG] Line height bug when dropping element before 'text' element
- [x]  [REFACTOR] Repeated code, especially in Block.tsx
- [ ]  [STYLE] Chevron animation on block description toggle
- [ ]  [STYLE] Improved block description CSS

### CODE REVIEW

**Comment by MushkyQT** [comment]
Closing to re-submit as smaller, more complete (functionality wise) PRs

**Comment by MushkyQT** [comment]
Nevermind 😅 
All major bugs should be fixed now.

File: src/features/blocks/Block.tsx [hunk 1]
```diff
@@ -1,51 +1,80 @@
...
 import {ToolboxCategoryBlocks} from "../../toolboxconfig";
 import {useEffect, useRef} from "react";
-import {useAppDispatch} from "../../hooks";
+import {useAppDispatch, useAppSelector} from "../../hooks";
 import {toggleBlockDescriptionAction} from "../../store/features/blocks/blocks";
+import {Draggable} from "react-beautiful-dnd";
+import {getDragStyle} from "../../App";
 
 function Block(props: ToolboxCategoryBlocks) {
+  const cat = useAppSelector(state => state.blocksReducer.categories.find((c) => c.blocks.find((b) => b.id === props.id)))
```
Comment (SebastienTainon) [inline src/features/blocks/Block.tsx]: You don't need the parenthesis when there is only one argument to the lambda function, eg:

`elements.find(c => condition(c))`

File: src/features/blocks/Block.tsx [hunk 2]
```diff
@@ -1,51 +1,80 @@
...
         } else {
           maxHeight = 0
         }
         blockDescriptionRef.current.style.maxHeight = maxHeight + "px"
       }
     })()
-  }, [props.toggled])
+  }, [cat, props.id])
+
+  enum TagTypes {
```
Comment (SebastienTainon) [inline src/features/blocks/Block.tsx]: Why repeating this? You could just reuse the TagType you've exported from `editorconfig.ts`?

File: src/features/blocks/Block.tsx [hunk 3]
```diff
@@ -1,51 +1,80 @@
...
+    Opening = 'opening',
+    Closing = 'closing'
+  }
+
+  function makeToolboxDraggable(tagProp: string, type: TagTypes, index: number) {
+    let classesToAdd = type === 'opening' ? 'toolbox-block-tag tag-open' : 'toolbox-block-tag tag-close'
```
Comment (SebastienTainon) [inline src/features/blocks/Block.tsx]: TagType.Opening instead of 'opening' :wink:

File: src/features/blocks/BlocksCategory.tsx [hunk 4]
```diff
@@ -32,23 +23,22 @@ function BlocksCategory(props: ToolboxCategory) {
         blocksContainer.style.maxHeight = maxHeight + "px"
       }
     })()
-  }, [props.toggled, props.blocks])
+  }, [open, props.openDesc])
```
Comment (SebastienTainon) [inline src/features/blocks/BlocksCategory.tsx:26]: Nicely done! If it doesn't need to be in the Redux state, using a local component state is better :+1:

File: src/features/editors/VisualHTMLEditorLine.tsx [hunk 5]
```diff
@@ -11,7 +11,8 @@ interface ILine {
 
 function Line(props: ILine) {
   const makeTag = (tag: LineSegment) => tag.type !== 'text' ?
-    tag.type === 'closing' ? `</${tag.value}>` : `<${tag.value}>` :
+    tag.type === 'closing' ? `</${tag.value}>` : `<${tag.value}>`
```
Comment (SebastienTainon) [inline src/features/editors/VisualHTMLEditorLine.tsx]: Idem, use TagType enum

File: src/store/features/editors/visualHTML.ts [hunk 6]
```diff
@@ -13,14 +17,34 @@ type MoveElement = {
   payload: DropResult
 }
 
+type DeleteElement = {
+  type: typeof EditorActionsTypes.EditorElementDelete
```
Comment (SebastienTainon) [inline src/store/features/editors/visualHTML.ts]: You don't need the `typeof`, it's already a type :wink:

**Review by SebastienTainon — CHANGES_REQUESTED** [review]
Btw, it's not directly related to this PR but I'm considering moving to Redux Toolkit on the Codecast project. The current Codecast architecture is not easy to manipulate, and dates from way before RTK release & adoption. I'll work on a POC this week, and I'll update you as soon so that you can also use it as soon as it's decided :wink: 

Sorry for the delay on this PR!

File: src/features/blocks/Block.tsx (hunk 2 above)
Comment (MushkyQT) [reply src/features/blocks/Block.tsx]: Not sure how this happened 😬 Thanks for catching it!

File: src/features/blocks/Block.tsx (hunk 3 above)
Comment (MushkyQT) [reply src/features/blocks/Block.tsx]: Fixed, thanks!

File: src/features/blocks/BlocksCategory.tsx (hunk 4 above)
Comment (MushkyQT) [reply src/features/blocks/BlocksCategory.tsx:26]: Thank you & roger that!

File: src/store/features/editors/visualHTML.ts (hunk 6 above)
Comment (MushkyQT) [reply src/store/features/editors/visualHTML.ts]: Yes I see, thanks for pointing this out!

**Pull request merged**